File: scripts/bug_hunt.py
#!/usr/bin/env python3
"""Bug Hunter: use Lean spec as oracle to find Rust implementation bugs."""

import subprocess, os, sys

NEVE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

def run_rust(source, timeout=30):
    tmp = os.path.join(NEVE_DIR, "tmp_bug.neve")
    with open(tmp, "w") as f: f.write(source)
    neve_bin = os.path.join(NEVE_DIR, "target", "release", "neve")
    if os.path.exists(neve_bin):
        r = subprocess.run([neve_bin, "run", tmp],
            capture_output=True, text=True, cwd=NEVE_DIR, timeout=timeout)
    else:
        r = subprocess.run(["cargo", "run", "-q", "-p", "neve", "--", "run", tmp],
            capture_output=True, text=True, cwd=NEVE_DIR, timeout=timeout)
    os.remove(tmp)
    return r.stdout.strip(), r.stderr.strip(), r.returncode

def check(name, src, expected, desc, bugs):
    stdout, stderr, code = run_rust(src)
    result = stdout.strip().lower()
    exp = str(expected).lower()
    ok = (exp in result) or (result == exp)
    if ok:
        print(f"  [PASS] {name}: {desc}")
    else:
        bugs.append((name, expected, result, desc, stderr[:200]))
        print(f"  [BUG]  {name}: expected='{expected}', got='{result[:60]}'")

def main():
    bugs = []
    print("=" * 55)
    print("  Bug Hunt: stdin limits (H-1)")
    print("=" * 55)
    check("stdin-ok", 'let c=io.commandWith(#{program="cat",args=[],stdin="hi"});let r=io.execCommand(c);println(io.processSuccess(r))', "true", "small stdin works", bugs)
    check("no-stdin", 'let r=io.execCommand(io.command("echo",["hello"]));println(io.processSuccess(r))', "true", "no stdin works", bugs)
    check("pipeline-stdin", 'let c=io.commandWith(#{program="cat",args=[],stdin="pipe-in"});let p=io.pipeline([c,io.command("cat",[])]);let r=io.execPipeline(p);println(io.processStdout(r))', "pipe-in", "pipeline stdin flows", bugs)

    print("\n" + "=" * 55)
    print("  Bug Hunt: output capture (H-2)")
    print("=" * 55)
    check("capture-stdout", 'let r=io.execCommand(io.command("echo",["hello world"]));println(io.processStdout(r))', "hello world", "stdout captured", bugs)
    check("capture-stderr", 'let r=io.execCommand(io.command("sh",["-c","echo err>&2"]));println(io.processStderr(r))', "err", "stderr captured", bugs)
    check("exit-code-0", 'let r=io.execCommand(io.command("true",[]));println(io.processSuccess(r))', "true", "exit 0 is success", bugs)
    check("exit-code-1", 'let r=io.execCommand(io.command("false",[]));println(io.processSuccess(r))', "false", "exit 1 is failure", bugs)

    print("\n" + "=" * 55)
    print("  Bug Hunt: env filtering (M-4)")
    print("=" * 55)
    check("safe-env", 'let c=io.commandWith(#{program="sh",args=["-c","echo $FOO"],env=#{FOO="bar"}});let r=io.execCommand(c);println(io.processStdout(r))', "bar", "safe env passes", bugs)
    # Test LD_PRELOAD stripping: use test -z to check if var is empty
    check("ld-preload-stripped", 'let c=io.commandWith(#{program="sh",args=["-c","test -z $LD_PRELOAD && echo stripped || echo LEAK"],env=#{LD_PRELOAD="/tmp/evil.so"}});let r=io.execCommand(c);println(io.processStdout(r))', "stripped", "LD_PRELOAD stripped from child env", bugs)

    print("\n" + "=" * 55)
    print("  Bug Hunt: spawn/await lifecycle")
    print("=" * 55)
    check("spawn-await", 'let c=io.command("echo",["spawned"]);let t=io.taskCommand(c);let r=io.awaitTask(t);println(io.processStdout(r))', "spawned", "spawn/await works", bugs)

    print(f"\n{'='*55}")
    if bugs:
        print(f"  {len(bugs)} POTENTIAL BUGS FOUND:")
        for name, exp, got, desc, _ in bugs:
            print(f"    {name}: {desc}")
            print(f"    expected={exp}, got={got}")
        print(f"{'='*55}")
        return 1
    else:
        print(f"  All tests passed. No bugs found.")
        print(f"{'='*55}")
        return 0

if __name__ == "__main__":
    sys.exit(main())

File: neve-cli/examples/bug_hunt.rs
/// Run with: cargo run --example bug_hunt -- [args]
// Usage: ./scripts/bug-hunt.rs
//
// Runs security regression tests by executing neve scripts and
// verifying stdout/stderr/exit codes against expected values.
use std::process::{Command, exit};

struct TestCase {
    name: &'static str,
    source: &'static str,
    expected: &'static str,
    desc: &'static str,
}

fn run_neve(source: &str) -> (String, String, i32) {
    let tmp = std::env::temp_dir().join(format!("neve_bug_hunt_{}.neve", std::process::id()));
    std::fs::write(&tmp, source).expect("write tempfile");

    let output = Command::new("cargo")
        .args(["run", "-q", "-p", "neve", "--", "run"])
        .arg(&tmp)
        .output()
        .expect("failed to execute neve");

    let _ = std::fs::remove_file(&tmp);

    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
    let code = output.status.code().unwrap_or(-1);

    // Combine stdout and stderr for matching (neve uses print/println)
    let combined = if stdout.is_empty() {
        stderr.clone()
    } else {
        stdout
    };
    (combined, stderr, code)
}

fn check(test: &TestCase, bugs: &mut Vec<String>) {
    print!("  [{}] ... ", test.name);
    let (output, stderr, code) = run_neve(test.source);

    let out_lower = output.to_lowercase();
    let exp_lower = test.expected.to_lowercase();
    let ok = out_lower.contains(&exp_lower) || out_lower == exp_lower;

    if ok {
        println!("PASS  {}", test.desc);
    } else {
        let msg = format!(
            "BUG   {}: expected='{}', got='{}', stderr='{}'",
            test.name,
            test.expected,
            &output[..output.len().min(80)],
            &stderr[..stderr.len().min(200)]
        );
        bugs.push(msg);
        println!(
            "FAIL  {} — expected='{}', got='{}'",
            test.desc,
            test.expected,
            &output[..output.len().min(60)]
        );
    }
    let _ = code;
}

fn main() {
    let tests = vec![
        // H-1: stdin limits
        TestCase {
            name: "stdin-ok",
            source: r#"
import std.io as io;
let cmd = io.commandWith("cat", [], cwd=None, stdin="hi", env=#{});
let r = io.execCommand(cmd);
io.processStdout(r)
"#,
            expected: "hi",
            desc: "small stdin works",
        },
        TestCase {
            name: "no-stdin",
            source: r#"
import std.io as io;
let r = io.execCommand(io.command("echo", ["hello"]));
io.processStdout(r)
"#,
            expected: "hello",
            desc: "no stdin works",
        },
        // H-2: output capture
        TestCase {
            name: "capture-stdout",
            source: r#"
import std.io as io;
let r = io.execCommand(io.command("echo", ["hello world"]));
io.processStdout(r)
"#,
            expected: "hello world",
            desc: "stdout captured",
        },
        TestCase {
            name: "capture-stderr",
            source: r#"
import std.io as io;
let r = io.execCommand(io.command("sh", ["-c", "echo err>&2"]));
io.processStderr(r)
"#,
            expected: "err",
            desc: "stderr captured",
        },
        TestCase {
            name: "exit-code-0",
            source: r#"
import std.io as io;
let r = io.execCommand(io.command("true", []));
toString(io.processSuccess(r))
"#,
            expected: "true",
            desc: "exit 0 is success",
        },
        TestCase {
            name: "exit-code-1",
            source: r#"
import std.io as io;
let r = io.execCommand(io.command("false", []));
toString(io.processSuccess(r))
"#,
            expected: "false",
            desc: "exit 1 is failure",
        },
        // M-4: env filtering
        TestCase {
            name: "safe-env",
            source: r#"
import std.io as io;
let cmd = io.commandWith("sh", ["-c", "echo $FOO"], cwd=None, stdin=None, env=#{FOO="bar"});
let r = io.execCommand(cmd);
io.processStdout(r)
"#,
            expected: "bar",
            desc: "safe env passes",
        },
        TestCase {
            name: "ld-preload-stripped",
            source: r#"
import std.io as io;
let cmd = io.commandWith("sh", ["-c", "test -z $LD_PRELOAD && echo stripped || echo LEAK"], cwd=None, stdin=None, env=#{LD_PRELOAD="/tmp/evil.so"});
let r = io.execCommand(cmd);
io.processStdout(r)
"#,
            expected: "stripped",
            desc: "LD_PRELOAD stripped from child env",
        },
        // spawn/await lifecycle
        TestCase {
            name: "spawn-await",
            source: r#"
import std.io as io;
let c = io.command("echo", ["spawned"]);
let t = io.taskCommand(c);
let r = io.awaitTask(t);
io.processStdout(r)
"#,
            expected: "spawned",
            desc: "spawn/await works",
        },
    ];

    println!("=== Bug Hunt: Security Regression Tests (Rust) ===");
    println!();

    let mut bugs = Vec::new();
    for test in &tests {
        check(test, &mut bugs);
    }

    println!();
    if bugs.is_empty() {
        println!("All {} tests passed. No bugs found.", tests.len());
    } else {
        println!("{} POTENTIAL BUGS FOUND:", bugs.len());
        for bug in &bugs {
            println!("  {bug}");
        }
        exit(1);
    }
}
